src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs

```diff
@@ -167,30 +179,40 @@ internal static void RecycleSocketAsyncEventArgs(SocketAsyncEventArgs args)
                 }
             }
         }
-        /// <summary>
-        /// Releases all resources owned by the connection
-        /// </summary>
-        protected virtual void Dispose(bool disposing)
+
+        public async Task DisposeAsync()
         {
-            if (disposing)
+            if (!_disposed)
             {
+                _disposed = true;
                 _stopping = true;
                 _output.Reader.CancelPendingRead();
 
-                Task.WaitAll(_sendTask, _receiveTask);
+                await Task.WhenAll(_sendTask, _receiveTask);
```

Comment:
As mentioned in https://github.com/dotnet/corefxlab/pull/1131#issuecomment-275172475 you could turned it into a `WhenAll` continuation? Not sure you really care about when it finishes with all the `async void` about..?